ambari-web/app/mappers/stack_service_mapper.js

```diff
@@ -119,7 +119,8 @@ App.stackServiceMapper = App.QuickDataMapper.create({
         }
         stackServiceComponents.push(parsedResult);
       }, this);
-      stackService.stack_id = stackService.stack_name + '-' + stackService.stack_version;
+      stackService.stack_id = `${stackService.stack_name}-${stackService.stack_version}`;
+      stackService.id = `${stackService.service_name}-${stackService.stack_id}`;
       stackService.service_components = serviceComponents;
```

Comment:
This will likely break many parts of the UI since code at many places assumes the value of id as service_name. Either we make all the other required changes or rename this as stackServiceId or something

Reply:
You are correct. This will break many things. But that is going to be the story of v3.0, since we are making fundamental changes to the way Ambari works and overturning long-held assumptions made in the codebase. I have actually already started trying to fix all of the references to StackService, but it definitely won't be perfect, since I don't have the ability to test all the areas touched by these changes.

Your suggestion to create separate property does not work because the problem is the ID property itself. The ID property is what Ember Data uses as its primary key and that was the root of this problem. When two mpacks have a service with the same name, and the service name by itself was being used as the primary key value, only one instance of the service would get added to the StackService store (since the second one has the same key, it was either not getting added or overwriting the first one) and thus only one service group would get the service installed.

Reply:
I agree. I meant only creating a separate property so that we don't break anything for now specially when we are nearing a demo. Of course in the long term the entire stack service model dependence would need be replaced as services are now mapped to mpacks and not stacks. But this change is breaking the dashboard itself. For eg: Mpack name and version in the service summary page is getting returned as null. So my suggestion is to make all the other appropriate changes before replacing this 'id' if you are against using a new property to represent this.

Reply:
Also we have other ways to fetch ember data and not necessarily by id. We can always fetch it using any other property for eg: App.StackService.find().findProperty('<property_name>', value)

Reply:
You don't understand. The problem is not with *retrieving* the item from the data store. The problem is with *storing* it in the data store. From what I can tell, the Ember Data stores use the ID as the primary key, and therefore do allow more than one record with the same ID to be stored. This is exactly how I would expect it to behave. However, *our data* has fundamentally changed so that *service name is no longer a unique identifier*. Therefore, we cannot store more than one record with the same service name if we continue to use service name as the value of ID. We *have to* use an ID value that is actually unique to actually store more than one record for the same service from different mpacks. Creating a separate property does not solve the problem at all.